tao_compiler/mlir/disc/transforms/disc_convert_fake_quant_op.cc

```diff
@@ -0,0 +1,169 @@
+    if (!inputFakeQuantOp || !weightFakeQuantOp) return failure();
+    if (inputFakeQuantOp.use_signed() != weightFakeQuantOp.use_signed() ||
+        inputFakeQuantOp.num_bits() != weightFakeQuantOp.num_bits()) {
+      return failure();
+    }
+    if (inputFakeQuantOp.use_dynamic() != op.use_dynamic() ||
+        weightFakeQuantOp.use_dynamic() != op.use_dynamic())
+      return failure();
+
+    Location loc = op.getLoc();
+    auto buildQuantizedTensorType = [&](mhlo_disc::FakeQuantOp fakeQuantOp) {
+      Type quantizedElemType =
+          fakeQuantOp.use_signed()
+              ? IntegerType::get(rewriter.getContext(), fakeQuantOp.num_bits())
+              : IntegerType::get(
+                    rewriter.getContext(), fakeQuantOp.num_bits(),
+                    mlir::IntegerType::SignednessSemantics::Unsigned);
+      auto inputTy = fakeQuantOp.input().getType().cast<RankedTensorType>();
+      return RankedTensorType::get(inputTy.getShape(), quantizedElemType,
+                                   inputTy.getEncoding());
+    };
+
+    // Builds a real quantized op by extracting metadata info from the
+    // fake_quant_op
+    auto buildQuantizedOpFromFakeQuantOp =
+        [&](mhlo_disc::FakeQuantOp fakeQuantOp) {
+          auto outTy = buildQuantizedTensorType(fakeQuantOp);
+          Value quantizedInput = rewriter.create<mhlo_disc::QuantizeOp>(
+              loc, outTy, fakeQuantOp.input(), fakeQuantOp.scale(),
+              fakeQuantOp.zero_point(), fakeQuantOp.use_symmetric(),
+              fakeQuantOp.axis(), fakeQuantOp.quant_min(),
+              fakeQuantOp.quant_max(), fakeQuantOp.use_dynamic());
+          return quantizedInput;
+        };
+
+    Value quantizedInput = buildQuantizedOpFromFakeQuantOp(inputFakeQuantOp);
+    Value quantizedWeight = buildQuantizedOpFromFakeQuantOp(weightFakeQuantOp);
+    Value quantizedConv = rewriter.create<mhlo_disc::QuantizedDynamicConvOp>(
+        loc, buildQuantizedTensorType(op), quantizedInput, quantizedWeight,
+        convOp.d_padding(), inputFakeQuantOp.scale(),
+        inputFakeQuantOp.zero_point(), weightFakeQuantOp.scale(),
+        weightFakeQuantOp.zero_point(), op.scale(), op.zero_point(),
+        *convOp.window_strides(), *convOp.padding(), *convOp.lhs_dilation(),
+        *convOp.rhs_dilation(), *convOp.window_reversal(),
+        convOp.dimension_numbers(), convOp.feature_group_count(),
+        convOp.batch_group_count(), op.use_symmetric(),
+        weightFakeQuantOp.axis(), op.use_dynamic());
+
+    Value dequantizedOut = rewriter.create<mhlo_disc::DequantizeOp>(
+        loc, op.getType(), quantizedConv, op.scale(), op.zero_point(),
+        op.use_symmetric(), op.axis(), op.use_dynamic());
+
+    rewriter.replaceOp(op, dequantizedOut);
+    return success();
+  }
+};
+
+void populateQuantizedPatterns(RewritePatternSet& patterns) {
+  // clang-format off
+  patterns.insert<
+    QuantizedDynamicConvOpConverter
```

Comment:
So there will be a quntized gemm converter? I see the document is demonstrating GEMM.

Reply:
Yes. I leave the quantized GEMM for you ^v^~